### PR TITLE
Reset session in/out pointers in mbedtls_ssl_session_reset()

### LINKED ISSUE #1941: session_in/out not reset in mbedtls_ssl_session_reset()
__Problem:__ `mbedtls_ssl_session_reset()`, specifically the underlying `ssl_session_reset_int()`, does not reset the pointers `mbedtls_ssl_context::session_in` and `mbedtls_ssl_context:session_out`, but it does free and `NULL`ify the `mbedtls_ssl_context::session` to which they map after a successful handshake. This leaves them dangling and e.g. makes a subsequent call to `mbedtls_ssl_get_max_frag_len()` fail with a segmentation fault (see [here](https://github.com/ARMmbed/mbedtls/blob/development/library/ssl_tls.c#L6887) for the failing access). 

__Further accesses:__ There is another entirely unrestricted [access](https://github.com/ARMmbed/mbedtls/blob/development/library/ssl_tls.c#L6840) to `session_out` in `mbedtls_ssl_get_record_expansion()`, i.e. not even guarded by a check that `session_out != NULL` (moving the subsequent check `transform_out != NULL` prior to the access would solve the matter).
Apart from that, there don't seem to be further critical accesses to `session_in/out`: There is an [access](https://github.com/ARMmbed/mbedtls/blob/development/library/ssl_tls.c#L1363) in `ssl_encrypt_buf()`, but this is guarded by `transform_out != NULL`, and `transform_out` _is_ properly reset in `mbedtls_ssl_sessoin_reset()`. The same holds for the [access](https://github.com/ARMmbed/mbedtls/blob/development/library/ssl_tls.c#L3033) in `mbedtls_ssl_write_record()`. There are similar accesses to `session_in` in `ssl_decrypt_buf()` and `ssl_prepare_record_content()` which are non-critical because they are guarded by a check that `transform_in != NULL`.

### PR DESCRIPTION
**Pull request by hanno-arm**
Fixes #1941.

### CODE REVIEW

File: ChangeLog [hunk 1]
```diff
@@ -10,6 +10,8 @@ Bugfix
    * Add ecc extensions only if an ecc based ciphersuite is used.
      This improves compliance to RFC 4492, and as a result, solves
      interoperability issues with BouncyCastle. Raised by milenamil in #1157.
+   * Fix potential segmentation fault in mbedtls_ssl_get_max_frag_len()
```
Comment (gilles-peskine-arm) [inline ChangeLog]: Why would there be a segmentation fault? Please be more specific. Null pointer dereference?

Comment (hanno-arm) [reply ChangeLog]: @gilles-peskine-arm Ok, it would be more precise to say that it's a potential use-after-free.

Comment (mpg) [reply ChangeLog]: Now that Gilles mentions it, I'm thinking perhaps we should always avoid the wording "segmentation fault" regardless of the cause, as it's a platform-specific behaviour.

**Comment by mpg** [comment]
@hanno-arm Note: #1939 is now based on this. In order to avoid unnecessary complexity, please refrain from rewriting history in this PR from now on.

**Comment by mpg** [comment]
Note: commit https://github.com/ARMmbed/mbedtls/pull/1939/commits/19c62f90e4608fc57f382cdbe8799ffdb98c9dda in #1939 adds a non-regression test for this.

**Pull request merged**